tools/tiller-hack/tiller-hack.go

```diff
@@ -0,0 +1,256 @@
+	}
+	bckpCfgMap := v1.ConfigMap{}
+	cfgMap.DeepCopyInto(&bckpCfgMap)
+	bckpCfgMap.ResourceVersion = ""
+
+	timestamp := strings.Replace(time.Now().Format("2006-01-0215:04:05.999"), "-", "", -1)
+	timestamp = strings.Replace(timestamp, ":", "", -1)
+	bckpCfgMap.ObjectMeta.Name += "." + timestamp + ".backup"
+	_, err = k8sClient.Core().ConfigMaps(configMapNamespace).Create(&bckpCfgMap)
+
+	return err
+}
+
+func decodeRelease(data []byte) (*rspb.Release, error) {
+	// base64 decode string
+	b, err := base64.StdEncoding.DecodeString(string(data))
+	if err != nil {
+		return nil, err
+	}
+
+	if bytes.Equal(b[0:3], magicGzip) {
+		r, err := gzip.NewReader(bytes.NewReader(b))
+		if err != nil {
+			return nil, err
+		}
+		b2, err := ioutil.ReadAll(r)
+		if err != nil {
+			return nil, err
+		}
+		b = b2
+	}
+
+	var rls rspb.Release
+	// unmarshal protobuf bytes
+	if err := proto.Unmarshal(b, &rls); err != nil {
+		return nil, err
+	}
+	return &rls, nil
+}
+
+func encodeRelease(rls *rspb.Release) (string, error) {
+	b, err := proto.Marshal(rls)
+	if err != nil {
+		return "", err
+	}
+	var buf bytes.Buffer
+	w, err := gzip.NewWriterLevel(&buf, gzip.BestCompression)
+	if err != nil {
+		return "", err
+	}
+	if _, err = w.Write(b); err != nil {
+		return "", err
+	}
+	w.Close()
+
+	return base64.StdEncoding.EncodeToString(buf.Bytes()), nil
+}
+
+func updateConfigMap(k8sClient *kubernetes.Clientset, configMapName, configMapNamespace, updatedData string) error {
+	cfgMap, err := k8sClient.Core().ConfigMaps(configMapNamespace).Get(configMapName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("fail to get configmap %s/%s with error: %+v", configMapNamespace, configMapName, err)
+	}
+	cfgMap.Data["release"] = updatedData
+
+	err = retry.RetryOnConflict(retry.DefaultBackoff, func() (err error) {
+		_, err = k8sClient.CoreV1().ConfigMaps(configMapNamespace).Update(cfgMap)
+		return
+	})
+	return err
+}
+
+func restartTiller(k8sClient *kubernetes.Clientset, namespace string) error {
+	selector := labels.SelectorFromSet(labels.Set(map[string]string{"app": "helm", "name": "tiller"}))
+	options := metav1.ListOptions{LabelSelector: selector.String()}
+	pods, err := k8sClient.Core().Pods(namespace).List(options)
+	if err != nil {
+		return err
+	}
+	if len(pods.Items) != 1 {
+		return fmt.Errorf("got unexpected number of tiller pod, failing")
+	}
+	var gracePeriod int64
+	return k8sClient.Core().Pods(namespace).Delete(pods.Items[0].ObjectMeta.Name,
+		&metav1.DeleteOptions{GracePeriodSeconds: &gracePeriod})
+}
+
+func main() {
+	flag.Parse()
+	flag.Set("logtostderr", "true")
+
+	k8sClient, err := BuildClient(*kubeconfig)
+	if err != nil {
+		glog.Errorf("Failed to create a client: %+v\n", err)
+		os.Exit(1)
+	}
+	if *configMapName == "" {
+		glog.Errorf("configmap name is missing.")
+		os.Exit(1)
+	}
+	if *configMapNamespace == "" {
+		*configMapNamespace = v1.NamespaceDefault
```

Comment:
this is the wrong way to set a default flag field.

Reply:
it is actually a very right way to specify default namespace, but not in the logic of this code. SO removing it.